refactor: Fetch and cache GutenbergKit editor settings

dcalhoun opened the pull request:
Fetch and cache editor settings before providing them to the editor, as it
allows improving offline support. The changes also communicate the 
background activity with an activity indicator. 

Ref CMM-200.

Related: 

* CMM-200-linear-issue
* https://github.com/wordpress-mobile/GutenbergKit/issues/105
* https://github.com/wordpress-mobile/GutenbergKit/pull/114
* https://github.com/wordpress-mobile/WordPress-Android/pull/21791

To test:

> [!Important]
> Ensure the following experimental features are enabled:
> * Experimental block editor
> * Experimental block editor styles

**1: Initial editor fallbacks to default settings**
1. Enable airplane mode. 
2. Open the post editor. 
3. Verify the editor loads, using default settings—e.g., white background with black text. 

**2: Editor fetches site-specific settings**
1. Disable airplane mode. 
2. Open the post editor. 
3. Verify the editor loads with site-specific settings—e.g., the theme's background and foreground colors. 

**3: Editor fallback to cached settings**
1. Enable airplane mode. 
2. Open the post editor. 
3. Verify the editor loads with site-specific settings—e.g., the theme's background and foreground colors. 


## Regression Notes
1. Potential unintended areas of impact
  Regressions in Gutenberg Mobile or Aztec editors. 
1. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manually smoke tested the editors. 
1. What automated tests I added (or what prevented me from doing so)
  Deemed unnecessary for the experimental editor. 
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dcalhoun committed:
hoist settings

dcalhoun committed:
refactor: Update settings via JavaScript bridge

dcalhoun committed:
fix: Fetch the entirety of the editor settings

dcalhoun committed:
Revert "refactor: Update settings via JavaScript bridge"

This reverts commit e4ec71b6c77e2439e9d98c0273f9bdcff64d098f.

dcalhoun committed:
feat: Postpone editor initialization

dcalhoun committed:
Revert "feat: Postpone editor initialization"

This reverts commit 39ac9caf71089853e5f256697f4882cc55d2bffd.

dcalhoun committed:
refactor: Defer editor start

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
refactor: Simplify RawBlockEditorSettingsService

Remove unnecessary completion handler structure.

dcalhoun committed:
feat: Cache editor settings fetch

Improve editor start performance.

dcalhoun committed:
feat: Editor settings use stale-while-revalidate cache

Mitigate stale editor settings.

dcalhoun committed:
style: Use shorthand

dcalhoun committed:
feat: Cache editor settings across editor sessions

The cache was discarded each time the editor closed, which defeats the
purpose of the cache. We must store it somewhere outside the editor view
controller.

dcalhoun committed:
refactor: Avoid duplicative settings requests

Avoid sending both foreground and background requests.

dcalhoun committed:
feat: Remove time-based cache

Revalidating the cache on each editor launch will ensure the latest
settings are available.

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
build: Update GutenbergKit ref

jkmassel commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -311,6 +321,53 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
+
+    // MARK: - Block Editor Settings
+
+    private func fetchBlockEditorSettings() {
```
This is a good start, but do we need to wait until the editor is launched to fetch these settings?

I'm wondering if we could do it while the user is viewing the "site dashboard"? If the fetch is still in progress while loading the editor we can show the spinner, but if not the editor should load much more quickly

WDYT?

dcalhoun replied:
Agreed. We should perform this fetch much earlier. It could be that we do not actually need the "warm up" mode class property, as the warm up could fetch the editor settings as well. 

Currently, the warm up run does not have any of the site configuration, so it cannot fetch the editor settings. We'll need to update the logic to access things like the site API URL.

dcalhoun replied:
I opt to addressed this now in 7768be7a5599f3e01071a68125ae167e827d171b. Let me know if you spot improvements for the implementation.

jkmassel commented on WordPress/Classes/Services/RawBlockEditorSettingsService.swift:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @MainActor
+    func fetchSettings() async throws -> [String: Any] {
```
What happens if we've never fetched the settings and the network connection fails?

In my testing it _seems_ like it'll just show a spinner forever?

If we don't have locally cached settings and the network request fails we should probably:

1. Try again some number of times (maybe 2?)
2. Show an error message 

I could see an argument for showing _some_ kind of editor with no remote settings, but I don't know how feasible that is?

dcalhoun replied:
Your experience is surprising. These changes were built to accomplish what you describe. When I block the network request (via Charles Proxy) for the editor settings, the editor continues to load with default editor settings after a moment. 

How are you causing the network connection failure in your testing? 

The screen recording below showcases both the default settings and site-specific settings. 

https://github.com/user-attachments/assets/489279e2-3213-4468-8ad2-7df8ae858c1b

dcalhoun replied:
In fact, the situation you describe is a part of the testing instructions in the PR description. 😄 I'm perplexed as to why you may be experiencing a different outcome.

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
perf: Preload editor settings within My Site

Improve load speed of editor.

dcalhoun committed:
fix: Enable GutenbergKit warmup via remote feature flag

This conditional was overlooked when the local experimental feature
toggle was augmented with a remote feature flag for roll out control.

https://github.com/wordpress-mobile/WordPress-iOS/pull/24465

jkmassel commented:
So it's not _forever_, but it is almost 5 minutes:

`gutenbergkit-measure_editor-first-render: 298.35688507556915`

I can consistently repro this by enabling the Network Link Conditioner (on a real device), setting it to "Very Bad Network", then running this PR on that device and following the testing instructions.

I suspect we can resolve the issue by putting a timeout on how long we wait for the remote network settings (I'd suggest 5 seconds – IINM, the connection timing out should automatically cause the "whoops, we couldn't load editor settings" dialog.

dcalhoun committed:
feat: Start editor without settings after three seconds

Avoid postponing the editor start for longer than three seconds. We
instead present the editor without any site-specific settings.

dcalhoun committed:
refactor: DRY up fetching editor settings

jkmassel commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -316,6 +327,70 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
+
+        Task { @MainActor in
+            // Start the editor with default settings after 3 seconds
+            let timeoutTask = Task {
```
This is a clever way to address this – I think ideally we'd just use the `timeout` property on `URLRequest`, but we don't _really_ own `WordPressOrgRestApi`. I guess we'll come back to this when it's time to adopt the Rust-layer REST API, and we can switch it over then.

jkmassel reviewed (approved):
Having discussed this in Slack, it seems this is as good as it'll get for now. We can control the settings loading prior to the editor being displayed, but unfortunately there's no UI inside the editor to show the other network operations that are happening.

This looks solid, and should work well in the vast majority of cases – if someone's network is this bad they're probably having a hard time using the app anyway.

dcalhoun commented:
> [...] unfortunately there's no UI inside the editor to show the other network operations that are happening.

Long term, I believe we can improve the experience with WebView-driven UI so that it display a "skeleton" while additional requests complete. 

That said, it gets complicated for the remote/site-specific editor. That editor fetches its React library (used for composing UI)—do we await React or build skeleton UI without React? We likely do the latter, and we can also finish implementing [cached bundles](https://github.com/wordpress-mobile/GutenbergKit/pull/107) to further improve the loading experience.

dcalhoun committed:
build: Update GutenbergKit ref
